src/java/org/apache/cassandra/repair/ParentRepairSessionListener.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.repair;
+
+import java.util.UUID;
+import java.util.concurrent.Future;
+
+import org.apache.cassandra.config.CassandraRelevantProperties;
+import org.apache.cassandra.service.ActiveRepairService;
+
+public interface ParentRepairSessionListener
+{
+    ParentRepairSessionListener instance = CassandraRelevantProperties.REPAIR_PARENT_SESSION_LISTENER.isPresent()
+                                      ? make(CassandraRelevantProperties.REPAIR_PARENT_SESSION_LISTENER.getString())
+                                      : new NoopParentRepairSessionListener();
+
+    static ParentRepairSessionListener make(String customImpl)
+    {
+        try
+        {
+            return (ParentRepairSessionListener) Class.forName(customImpl).newInstance();
+        }
+        catch (Throwable ex)
+        {
+            throw new IllegalStateException("Unknown parent repair session listener: " + customImpl);
+        }
+    }
+
+    /**
+     * Call when parent repair session is registered
+     */
+    default void onRegistered(UUID sessionId, ActiveRepairService.ParentRepairSession session) {}
+
+    /**
+     * Call when parent repair session is removed
+     */
+    default void onRemoved(UUID sessionId, ActiveRepairService.ParentRepairSession session) {}
+
+    /**
+     * Call when validation task started for given repair session
+     */
+    default void onValidation(RepairJobDesc desc, Future validationTask) {}
+
+    /**
+     * Call when sync task started for given repair session
+     */
+    default void onSync(RepairJobDesc desc, Future syncTask) {}
+
+    static class NoopParentRepairSessionListener implements ParentRepairSessionListener
```

Comment:
I'm sorry, I wasn't clear apparently.
We're still relying on default implementation in the interface.

The problem with that is if, say, another such method is added to the interface the implementation will never know. Compiler check for unimplemented methods is effectively turned off. That's especially important since cndb implementation lives in another repository.

I'd rather we had no `default` implementation in the interface unless you believe an empty implementation is really the most common, and that my worry described above is overstated.

As it is now, there's not much difference between a named class or `new ParentRepairSessionListener() {}`

Reply:
good idea, removed default implimentations